marklogic-data-hub-spark-connector/src/main/java/com/marklogic/hub/spark/sql/sources/v2/writer/HubDataWriter.java

```diff
@@ -142,7 +146,20 @@ protected JsonNode determineIngestionEndpointParams(Map<String, String> options)
         }
 
         if (doesNotHaveApiPath) {
-            endpointParams.put("apiPath", "/data-hub/5/data-services/ingestion/bulkIngester.api");
+            String apiPath = "/data-hub/5/data-services/ingestion/bulkIngester.api";
+
+            if(hubClient.getModulesClient().newJSONDocumentManager().exists(apiPath) == null) {
+                IOUtil ioUtil = new IOUtil(this.hubClient.getModulesClient());
+                try {
+                    ioUtil.load("bulkIngester.api");
```

Comment:
We'll need to toss these files into the shadowJar, which is part of https://github.com/marklogic/marklogic-data-hub/pull/4695 .

Once it's in the jar, the Spring class ClassPathResource can be used to retrieve it, no need for IOUtil:

    new ClassPathResource("marklogic-data-hub-spark-connector/bulkIngester.api").getInputStream()

Reply:
@rjrudin . I am guessing - 
new ClassPathResource("marklogic-data-hub-spark-connector/bulkIngester.api").getInputStream() 
is a replacement for IOUtil.testFileToString(). 
There are other operations in IOUtil as well like - adding permissions and writing to data-hub-MODULES database. Please let me know if there is a method in data-hub that handles the permissions part. Will go ahead and replace them.

Reply:
@anu3990 @SameeraPriyathamTadikonda I think you'll want a separate method for building a DocumentWriteOperation for each endpoint module that needs to be written (it would be unusual, but it's possible that the connector only needs to write 1 or 2 of the 3 endpoint modules). And then another method will handle adding each DWO to a DocumentWriteSet, which can then be written to the modules database. I think all of that logic should be in HubDataWriterSource for now - there's no need for a separate utility class, the Java Client makes it pretty simple. We'd just want to avoid any duplication within HDWS.